### DISCUSSION TITLE
CDN ?

### DISCUSSION THREAD

**Discussion by stevespaw** [dyc3/opentogethertube#1481] [Q&A]
As I have been thinking about use case - I realized that I had not considered CDN. Can I connect this to a CDN, or is it using the CDN of the source, and just referencing the Current Playtime? I am still trying to get a grip on how this works. Thanks for a cool tool!

**Comment by dyc3** [comment]
If you give OTT a link to an HLS stream or mp4 video, it will always use that URL when telling clients where to grab the video from. OTT doesn't restream the video for clients, it tells the clients the URL to stream the video from.

```mermaid
flowchart LR
    OTT-->A[Client 0]
    OTT-->B[Client 1]
    OTT-->C[Client N]

    A --> Source
    B --> Source
    C --> Source
```

Could you elaborate a little bit more on your use case? Based on the issues you've filed, it sounds like you're trying to sync video between a bunch of clients for a movie premiere style event?

**Reply by stevespaw** [reply]
>  it sounds like you're trying to sync video between a bunch of clients for a movie premiere style event?

Yes that would be a main use case! I looks as if the CDN of the "source" handles the Video Distribution, correct?

**Reply by stevespaw** [reply]
On an MP4, That still comes from a source? Such as s3? OTT does not stream it directly?

**Reply by stevespaw** [reply]
The workload of OTT is the frequent ping of play position?

**Reply by dyc3** [reply]
That's correct, each client will stream the video directly from the source. Using a CDN as the video source should work perfectly fine.

However, there are some things to note about streaming mp4s naively (eg. not using HLS/DASH, just making normal HTTP request and getting the whole file back as the response):
- Some browsers will always start streaming the video from the beginning. This means if you have a user join the room halfway through an hour long video, they will need to buffer all 30 minutes before the video will play smoothly for them.
- If the video metadata atom is not at the very beginning of the mp4, it could take a very long time for OTT to parse the video and actually add it to the queue. This can be avoided by setting `-movflags faststart` in ffmpeg when transcoding the video.
- Browsers do not support the ac3 audio codec, and some browsers do not support the x265 video codec. See the ffmpeg commands at the bottom of the FAQ: https://github.com/dyc3/opentogethertube/discussions/830

**Reply by dyc3** [reply]
> The workload of OTT is the frequent ping of play position?

During playback OTT doesn't send any messages unless something happens in the room. If you play/pause the video, it'll broadcast a message to all the clients. But while the video is playing, no messages are sent.

**Reply by stevespaw** [reply]
This is great information! I know the MP4 issues, Is there any benefit to MP4, or is HLS the best way to do this, which is what I assume. Thanks.

**Reply by dyc3** [reply]
The benefit to using mp4s is that it's easier to set up. 

So yeah, I would recommend setting up HLS instead. It's currently a little bit better supported by OTT than DASH.

**Reply by stevespaw** [reply]
So once playback starts there is no further Sync verification? Just asking..

**Reply by dyc3** [reply]
Yeah. The client keeps track of when playback is started and enforces the playback position of the player to be within ~1 ish second of the "real" playback position tracked by the server. Generally though, clients stay in sync within around ~100-200ms at the most. This negates the need to constantly verify with the server that it's still in sync.

**Reply by stevespaw** [reply]
Do you think this could handle 2 million viewers in a room?
Serious question..

**Reply by dyc3** [reply]
uhhh... I have no idea. I don't have the hardware necessary to test that kind of capacity. However, I could probably whip up a k6 load test for you so that you can run it on your own hardware to make that assertion.

**Reply by stevespaw** [reply]
Love it that you are willing to try!!!  I'm am trying to assess features now to present some ideas. This is like 95% of what we have been discussing for a while, I just ran across it. I am down to the Timer question and a new API question. :-).

**Reply by stevespaw** [reply]
We do have a 2 million viewer possible project.

**Reply by dyc3** [reply]
The biggest problem I foresee with 2 million people is the user list. Currently it's not paginated so it would end up rendering 2 million entries in _everyone's browser_. On top of that, every new user joining the room would cause a message to be broadcast. So definitely some changes would need to be implemented:
1. user list pagination
2. and/or a way to disable the user list

**Reply by stevespaw** [reply]
For this application, just a way to disable the list would be perfect.
